### ISSUE TITLE
[JRI] Retrieving variables to Java using GNU/Linux works, but does not in Windows.

### ISSUE THREAD

**Issue by vlaze** [s-u/rJava#192]
### **TL;DR**
Whenever you try to get a processed (see below) variable from a REXP in JRI, results vary between operating systems: GNU/Linux will work properly; Windows 10 will throw an exception. Please refer to "**How does it work?**" to get the code.

GNU/Linux output (correct):

![Programa básico - Salida Linux](https://user-images.githubusercontent.com/16038211/60762178-727ff500-a05a-11e9-8b9f-42182f1377e6.PNG)

Windows 10 (wrong):

![Programa básico - Salida Windows](https://user-images.githubusercontent.com/16038211/60762182-81ff3e00-a05a-11e9-9219-eceea4af43c5.PNG)

### **What am I doing?**

I am writing a data analysis program in Java which relies on R's ```arules``` library to mine association rules. 

### **What do I want?**

My purpose is to store the rules into a String variable in Java so that I can process them later.

### **How does it work?**

1.	Runs the ```apriori``` algorithm with the transaction list and some necessary values passed as parameter.
2.	Reorders the generated association rules.
3.	Given that the association rules cannot be stored as a String straight away, they are written to the standard output with R's ```write``` method, captured and stored it in a variable. I have converted the association rules into a string variable. We return the string.

The code is the following:

    // Step 0
        Rutils.rengine.eval("library(arules)");
        Rutils.rengine.eval("data(Groceries)");

    // Step 1
        Rutils.rengine.eval("rules <- apriori (Groceries, parameter = list(supp = 0.001, conf = 0.5))");

    // Step 2
        Rutils.rengine.eval("orderedRules <- sort(rules, by = c('count', 'lift'), order = FALSE)\n");

    // Step 3
        REXP rexp = Rutils.rengine.eval("rulesAsString <- paste(capture.output(write(orderedRules, file = stdout(), sep = ',', quote = TRUE, row.names = FALSE, col.names = FALSE)), collapse='\\n')\n");
        System.out.println("Regular expression: " + rexp.asString());
        return rexp.asString();   

### **What´s wrong?**

Running the code in Linux Will work perfectly, but when I try to run it in Windows, I get the following error referring to the ```return``` line:

    Exception in thread "main" java.lang.NullPointerException

This is a common error I have whenever the R code generates a null result and passes it to Java. There's no way to syntax check the R code inside Java, so whenever it's wrong, this error message appears. 

However, when I run the R code in brackets in the R command line in Windows, it works flawlessly, so both the syntax and the data flow are OK.

### **Technical information**

 - In Linux, I am using Arch Linux with R 3.6.0 (29-04-26) with OpenJDK 10. 0.2 (2018-07-17).

 - In Windows, I am currently using Oracle's latest JDK release (12), but trying to run the program with OpenJDK 12 for Windows does not solve anything. R is 3.6.0 too.

- rJava is installed in both machines' R at its latest version.

 - All software is 64 bits.

 - The IDE used in both operating systems is IntelliJ IDEA 2019.1.3. Community version for Linux and Ultimate versión for Windows.

### **Screenshots**
These are screenshots of the configurations that will remain the same for every further version of the sample program, regardless the OS.

Linux run configuration:

[![Linux run configuration][1]][1]

Windows run configuration:

[![Windows run configuration][2]][2]
**How to reproduce**
The original R code is working regardless the platform. Install the ```arules``` package and run this script in an R command line:

```
library(arules)
data(Groceries)
rules <- apriori (Groceries, parameter = list(supp = 0.001, conf = 0.5))
orderedRules <- sort(rules, by = c('count', 'lift'), order = FALSE)
rulesAsString <- paste(capture.output(write(orderedRules, file = stdout(), sep = ',', quote = TRUE, row.names = FALSE, col.names = FALSE)), collapse='\n')
print(rulesAsString)
```
Or just copy the Java code I wrote before.

  [1]: https://i.stack.imgur.com/S5y0p.png
  [2]: https://i.stack.imgur.com/poXH8.png

**Comment by vlaze** [comment]
### UPDATE 07/07/2019
Trying to figure out what causes the problem so as to eliminate possibilities, I wrote a small Java program to diagnose several JRI functionalities. 

### The test
There are 3 big tests consisting of data communication from R to Java:

- **Simple REXP generation**. Create a sample string in R and pass it to Java.

- **stdout() printing and capturing**. This error always happens when I use a combination of `write`, `capture.output` and optionally `paste` (I have tried to remove the `paste` instruction and the results did not change anything.

- **Algorithm + Test 2**. Having checked test 2 is not the cause of the problems, let's make it more challenging: we are going to mine some association rules with `apriori()` and we are going to process them the same way we did with a variable in Test 2.

In addition, one can issue `eval` either against an R variable or a bare R instruction. Let's check this example extracted from the program I made. Pay attention that `c <-` is used and not used respectively:

`rexp = r.eval("paste(capture.output(write(orderedRules, file = stdout())), collapse='\\n')");`
`rexp = r.eval("c <- paste(capture.output(write(orderedRules, file = stdout())), collapse='\\n')");`

Both cases are taken into account in the program, making a total number of 6 tests to evaluate.

### The code
Change the extension to `.java` and it should work with any properly configured JRI version: [Main.txt](https://github.com/s-u/rJava/files/3365884/Main.txt)

Or just copy, paste and save as a `.java` file:
```
import org.rosuda.JRI.REXP;
import org.rosuda.JRI.Rengine;

import java.text.DecimalFormat;
import java.text.NumberFormat;

public class Main {
    public static void main(String args[]) {
        // Optional decimal styling. Remove if needed.
        NumberFormat formatter = new DecimalFormat("#0.00");

        // We are considering 3 'big' tests with 2 use cases each. 6 tests if we take them into account.
        int testPassed = run(6);
        System.out.println("\n\n\nThe program has passed " + testPassed + "/6 tests. Success rate: " + formatter.format((float)testPassed/(float)6*100) + "%.");
        System.exit(0);
    }

    private static int run(int numberOfTests) {
        Rengine r = new Rengine(new String[]{("--no-save")}, false, null);
        r.eval("library(arules)");
        r.eval("data(Groceries)");

        System.out.println("=================== [TEST 1] Data from R to Java. Printing REXP as String. ===================");
        try {
            System.out.println("\n(Use case 1: Print from a variable.)");
            System.out.println("Created REXP variable with value \"whatever string\".");
            REXP a = r.eval("a <- \"whatever string\""); // We store in a Java variable an R variable assignation.
            System.out.println("Value as REXP: " + a);
            System.out.println("Value as String: " + a.asString());
            System.out.println("Value as Integer: " + a.asInt());
        }
        catch (NullPointerException np) {
            System.out.println("[ /!\\ TEST 1, USE CASE 1 FAILED! /!\\ ]");
            np.getStackTrace();
            numberOfTests--;
        }

        try {
            System.out.println("\n(Use case 2: Print directly.)");
            System.out.println("Created REXP variable with value \"whatever string\".");
            REXP a2 = r.eval("\"whatever string\""); // We store in a Java variable an R command.
            System.out.println("Value as REXP: " + a2);
            System.out.println("Value as String: " + a2.asString());
            System.out.println("Value as Integer: " + a2.asInt());
            System.out.println("\n");
        }
        catch (NullPointerException np) {
            System.out.println("[ /!\\ TEST 1, USE CASE 2 FAILED! /!\\ ]");
            numberOfTests--;
        }


        System.out.println("\n\n\n=================== [TEST 2] Writing to stdout() in R, capturing the output and storing it into a REXP variable. ===================");

        System.out.println("\n(Use case 1: Print from a variable.)");
        try {
            System.out.println("Created REXP variable. Expected value is \"Testing standard output\".\n");
            REXP b = r.eval("b <- paste(capture.output(write(\"Testing standard output...\", file = stdout())), collapse='\\n')"); // We store in a Java variable an R variable assignation.
            System.out.println("Value as REXP: " + b);
            System.out.println("Value as String: " + b.asString());
            System.out.println("Value as Integer: " + b.asInt());
        }
        catch (NullPointerException np) {
            System.out.println("[ /!\\ TEST 2, USE CASE 1 FAILED! /!\\ ]");
            numberOfTests--;
        }


        System.out.println("\n(Use case 2: Print directly.)");
        try {
            System.out.println("Created REXP variable. Expected value is \"Testing standard output\".\n");
            REXP b2 = r.eval("paste(capture.output(write(\"Testing standard output...\", file = stdout())), collapse='\\n')"); // We store in a Java variable an R command.
            System.out.println("Value as REXP: " + b2);
            System.out.println("Value as String: " + b2.asString());
            System.out.println("Value as Integer: " + b2.asInt());
        }
        catch (NullPointerException np) {
            System.out.println("[ /!\\ TEST 2, USE CASE 2 FAILED! /!\\ ]");
            numberOfTests--;
        }


        System.out.println("\n\n\n=================== [TEST 3] Association rule mining, sorting, capturing from output and pasting. ===================");
        REXP rexp = null;
        try {
            System.out.println("\n(Use case 1: Print from a variable.)");
            System.out.print("Running apriori...");
            r.eval("rules <- apriori (Groceries, parameter = list(supp = 0.001, conf = 0.5))");
            System.out.println(" DONE");
            System.out.print("Sorting rules...");
            r.eval("orderedRules <- sort(rules, by = c('count', 'lift'), order = FALSE)");
            System.out.println(" DONE");
            System.out.println("Created REXP variable, capturing the output and storing the value. Expected value is a huge string of rules.\n");
            rexp = r.eval("c <- paste(capture.output(write(orderedRules, file = stdout())), collapse='\\n')");  // We store in a Java variable an R variable assignation.
            System.out.println("Value as REXP: " + rexp);
            System.out.println("Value as String: " + rexp.asString());
            System.out.println("Value as Integer: " + rexp.asInt());
        }
        catch (NullPointerException np) {
            System.out.println("[ /!\\ TEST 3, USE CASE 1 FAILED! /!\\ ]");
            numberOfTests--;
        }

        try {
            System.out.println("\n(Use case 2: Print directly.)");
            System.out.print("Running apriori...");
            r.eval("rules <- apriori (Groceries, parameter = list(supp = 0.001, conf = 0.5))");
            System.out.println(" DONE");
            System.out.print("Sorting rules...");
            r.eval("orderedRules <- sort(rules, by = c('count', 'lift'), order = FALSE)");
            System.out.println(" DONE");
            System.out.println("Created REXP variable, capturing the output and storing the value. Expected value is a huge string of rules.\n");
            REXP rexp2 = r.eval("paste(capture.output(write(orderedRules, file = stdout())), collapse='\\n')");  // We store in a Java variable an R variable assignation.
            System.out.println("Value as REXP: " + rexp2);
            System.out.println("Value as String: " + rexp2.asString());
            System.out.println("Value as Integer: " + rexp2.asInt());
        }
        catch (NullPointerException np) {
            System.out.println("[ /!\\ TEST 3, USE CASE 2 FAILED! /!\\ ]");
            numberOfTests--;
        }
        return numberOfTests;
    }
}
```



### Test results
A score is calculated while the test is running to provide with a success rate after it finishes. Results have, as expected, been better in Linux. This can help to confirm there is at least one difference in the execution flow of JRI. 

**Windows**
Complete console output: 
[Windows.txt](https://github.com/s-u/rJava/files/3365886/Windows.txt)

![Versión de Java - Windows](https://user-images.githubusercontent.com/16038211/60771123-63915500-a0e3-11e9-9d66-5daa0611ccf0.PNG)
![Versión de R - Windows](https://user-images.githubusercontent.com/16038211/60771128-7441cb00-a0e3-11e9-81ed-92d5197b5b2e.PNG)
![Salida - Windows](https://user-images.githubusercontent.com/16038211/60771115-4492c300-a0e3-11e9-9182-fa52e2128ab6.PNG)

**GNU/Linux**
Complete console output: 
[Linux.txt](https://github.com/s-u/rJava/files/3365887/Linux.txt)

![Versión de Java - Linux](https://user-images.githubusercontent.com/16038211/60771137-8f143f80-a0e3-11e9-85a8-1b1eeb54a40b.PNG)
![Versión de R - Linux](https://user-images.githubusercontent.com/16038211/60771139-91769980-a0e3-11e9-909b-cd3f600abbe3.PNG)
![Salida - Linux](https://user-images.githubusercontent.com/16038211/60771141-963b4d80-a0e3-11e9-9074-ad25e0cb9b76.jpeg)

**Comment by s-u** [comment]
You're likely getting an R error - always use `try({ ... }, silent=TRUE)` so the Java side can collect errors.
